### ISSUE TITLE
Support intra-doc links on cross-crate reexports

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/rust#65983]
Previously: https://github.com/rust-lang/rust/issues/60883, https://github.com/rust-lang/rust/issues/59833

This is the last major issue blocking the stabilization of intra doc links (https://github.com/rust-lang/rust/issues/43466)

Currently, if you use intra-doc links on some item:

```rust
// crate "foo"

/// This is not [Baz]
pub struct Bar;

pub struct Baz;
```

the docs will not turn up on any crate reexporting it

```rust
pub use foo::Bar;
```

The docs will have a bare `[Baz]` instead of the appropriate intra-doc link, because we resolve intra-doc links locally. (More examples in https://github.com/rust-lang/rust/issues/60883, https://github.com/rust-lang/rust/issues/59833)

------

This is tricky to fix; we throw away module resolution information before packaging up crate metadata, so it's not possible to "unseal" a crate and rerun the resolution within it. One solution would be to pre-detect intra doc links and store resolutions in metadata, but this would be expensive since it involves parsing everything as markdown unconditionally.

A better solution is to save resolution information in the metadata, as some kind of tree keyed to the module.

@eddyb's [comment](https://github.com/rust-lang/rust/pull/47046#issuecomment-354824520) outlines the correct approach to use here

> What you actually want to persist cross-crate is the scope information that lets you resolve names in a certain (module) scope, after the fact. It'd also allow not doing all the the resolver hacks.
> 
> And there's a very good reason to want it in the compiler, at least same-crate: we could (finally) tell if a type is in scope somewhere so we can actually print `Vec<_>` instead of `std::vec::Vec<_>` etc.

Basically, instead of throwing away resolver metadata, we should serialize it into some form and store it in the metadata, so that it can be reloaded as necessary.

**Comment by kinnison** [comment]
I'm interested in helping on this, but right now I'm not confident in the requisite parts of the compiler.  I will take a poke around, but please don't let that stop someone who actually knows what they're doing getting it done :D

**Comment by eddyb** [comment]
At this point I really wish rustdoc always took a set of crates' sources and never relied on metadata from `rlib`s.
It would make so many things easier, including this.

**Comment by kinnison** [comment]
Okay, I've spent a bit of time trying to find my way around the libraries which comprise the compiler, looking for things associated with the metadata and/or name resolution.  I'm *starting* to get a feel for things, but I'm still not certain I know quite what I'm trying to achieve.  Since this is `E-hard` I wasn't expecting to work it out right away fortunately, and I'm prepared to keep learning, but it'd be incredibly helpful if someone were prepared to help point me toward the right places to be learning about.  Once I feel like I've done enough basic learning, I'll probably post a rough outline of what work I *think* needs doing to complete this (kinda as an expansion of Manish's original description) and then ask for review of that.

@Manishearth or @eddyb I'm guessing would be best to ask for that review, so this is me pre-asking to see if the above approach would work for either of you?

**Comment by Manishearth** [comment]
This would work for me, I think @eddyb is far more familiar with some of the stuff here than I am.

librustc_resolve contains a resolution tree that effectively needs serializing, but ideally a reduced form of it.

**Comment by eddyb** [comment]
Probably the best way to do this would be in two steps:
1. come up with some data structure to place in `librustc`
    * it only needs to handle being from the same crate, to begin with
    * `librustc_resolve` would export this, and `librustdoc` would use it to resolve (some?) links, potentially removing the need for "late `Resolver` usage" in `librustdoc`
    * @petrochenkov is most likely to know how and what to exfiltrate from `librustc_resolve`, I can only guess what parts are relevant
2. once you have that, you can make `librustc_metadata` encode/decode it
    * the actual serialization is based on `#[derive]`s so all you really need to do is add the plumbing around that
    * this is the part I can actually help with

**Comment by Manishearth** [comment]
> come up with some data structure to place in librustc 

Basically needs to be a map of module DefId to all contents (imported and otherwise) by name.

Something like `HashMap<DefId, ModuleNames>`, `struct ModuleName {types: HashMap<Ident, DefId>, values: .., macros: ..}`

idk how heavy that would be, or how much work that would be to compute

**Comment by kinnison** [comment]
That all sounds super-useful.  I guess I need to start learning what a `DefId` is etc. properly :D

**Comment by eddyb** [comment]
@Manishearth Hmm, sounds like the way `ExportMap` works, except not just for exports. I thought `rustdoc` would need something more fine-grained than `mod`s, but if not that's great.

**Comment by Manishearth** [comment]
@eddyb precisely.

Also, currently we don't do inner imports anyway, the following will not work:

```rust
fn foo() {
   //! [baz]
   use bar::baz;
}
```

It's debatable if we _want_ that to work, and in that case we can use local resolution though i'd prefer if we only used one way of resolving.

Maybe we can just make ExportMap contain all the module level items, with a pub/priv tag?

Of course this doesn't 100% solve your other problem of being able to spew scoped diagnostics, but it's a step.

**Comment by kinnison** [comment]
I have been thinking quite hard about this and I wonder if it might be sufficient to detect any `[...]` sequence in a docstring (should be cheap to do) and then do the path resolution work inside the compiler rather than deferring to when rustdoc begins?  It would then be able to append (or prepend) reference resolution lines to the doc block, thereby not needing *any* changes to the serialisation or resolver etc.

The argument @Manishearth made in the original posting that full parsing of the markdown would be expensive wouldn't have to happen for every docstring, only ones which contain the `[` character, and we could blithely create reference resolution lines for any `[...]` in the text without consideration of markdown if we wanted to be even faster.

I appreciate this really wouldn't help @eddyb but it strikes me as a potentially much simpler pathway to completion of this task.

**Comment by Manishearth** [comment]
that ... could work!

i would prefer we do the harder thing, but this would let us ship.

However, note that this means we can no longer emit warnings for broken links since we're not sure which things are meant to be intra doc links without preparsing markdown

**Comment by kinnison** [comment]
Oh good point about the lint/warning.  Hmm.  I guess that we really do need that independent lookaside unless we start doing daft things around embedded nulls in the docs to separate input from autogenerated resolutions, and that would be *HORRIBLE*.  I'll continue to think about possibly simpler solutions while also trying to continue to learn enough to do the "right" thing instead.

**Comment by Manishearth** [comment]
We should see if we can close this out, this is basically the one thing blocking us from stabilizing intra-doc-links, and folks have been asking for this feature for years.

There are roughly three ways to implement this:

 - librustc_resolve emits a module-imports map -- a structure mapping each module to a list of its imports. this gets stored in the metadata, we consume this in rustdoc
 - the rustc parser actually parses markdown in doc comments, stashes away all the unresolved paths, and librustc_resolve fixes them up with def ids, which get stored in
 - rustdoc _itself_ has its own form of metadata, and when it does intra-doc-link resolution it stashes the metadata somewhere


The last option is the one that appeals to me most because it has zero perf impact, but I don't know how best to do it architecturally (perhaps just a json file somewhere? idk how best to invalidate it). The first option is probably the easiest.

@rust-lang/compiler @rust-lang/rustdoc do y'all have opinions on how best to do this? Do you think adding some form of metadata to rustdoc itself is doable?

**Comment by kinnison** [comment]
For anyone reading this, I ended up unable to continue with this work, so the job is still entirely open for anyone to take.

**Comment by nikomatsakis** [comment]
Huh, interesting. This feels related to our long-standing efforts to extact name resolution into a shareable code based that both rust-analyzer and rustc can make use of. Not that we've made much progress on that.

**Comment by Manishearth** [comment]
Yeah, I recall @eddyb mentioning this before in that context.

Would simply storing the module tree in a longer-term queryable metadata-able map be a  worthwhile endeavor?

**Comment by Manishearth** [comment]
@nikomatsakis would you or someone on the compiler team (@eddyb ?) be able to sketch out what such an approach would entail?

**Comment by jyn514** [comment]
If someone is willing to mentor, I'd be interested in working on this.

**Comment by Manishearth** [comment]
@eddyb would you have time to do this? 

@jyn514 the basic idea is that the resolver tree in rustc_resolve needs to be serialized to something persistent. It's possible that the best way to do this is to simply make it no longer a self-referential arena and have it use ids instead. Or have a way to copy it out into a serializeable datastructure that has less of the mid-resolution tracking stuff in it, one that is easy to query.

A simple serialized datastructure would be a map from module defids to names imported in the module.

**Comment by jyn514** [comment]
I tried serializing all modules last night and ran into lots of problems because of all the arena allocation. Would it be simpler to instead serialize `ExportMap`, as suggested earlier?

> Maybe we can just make ExportMap contain all the module level items, with a pub/priv tag?

I think that would hit all of rustdoc's needs since it only requires a name -> DefId lookup. I'm not sure what currently goes into an ExportMap - maybe `ResolverOutputs` is closer to what we need, since rustdoc can link to other crates? I think either of those would be much easier than serializing a full `ModuleData`.

**Comment by jyn514** [comment]
cc @petrochenkov since @eddyb said he knows the most about name resolution ^

**Comment by jyn514** [comment]
While investigating this, I found rustdoc will always resolve imports relative to the current module. This doesn't show up for exports within a crate because rustdoc doesn't generate docs for those:
![image](https://user-images.githubusercontent.com/23638587/83950542-7cc1ea80-a7f9-11ea-8ca8-443db10ea156.png)

It only generates docs for cross-crate re-exports:
![image](https://user-images.githubusercontent.com/23638587/83950550-877c7f80-a7f9-11ea-8e72-4d375c4e76f9.png)

@rust-lang/rustdoc Is that intended behavior? Why is this different for cross-crate exports, does it have something to do with `--no-deps`?

@Nemo157 mentioned

> it only does re-exports if the item is documented elsewhere, if `f` were not pub then `inner::f` would be the primary location for it and get the full docs

but that doesn't seem to explain the docs for `Rng` since those have a primary location in the `rand` crate.

**Comment by Manishearth** [comment]
> but that doesn't seem to explain the docs for Rng since those have a primary location in the rand crate.

Right, that rule is for same-crate reexports

> I tried serializing all modules last night and ran into lots of problems because of all the arena allocation. Would it be simpler to instead serialize `ExportMap`, as suggested earlier?

The serialization is not the hard part, you need to make a version of `ExportMap` that contains private items (or tweak `ExportMap` itself to contain both, if you can do that in a way that doesn't affect perf.).

The whole issue to be solved is that that arena datastructure needs to be converted to something else that we can store in metadata (i.e. that we can serialize cross-crate). That might be an ExportMap, that might be something else. If we can make it work by extending ExportMap, that works just fine.

**Comment by jyn514** [comment]
In https://github.com/jyn514/rust/tree/rustdoc-absolute-module I changed rustdoc to pass in the parent module for all items, not just local items. That changes the docs for the following code:

```rust
/// This is documentation for the re-export of [`Rng`]
pub use rand::Rng;
```
from a dead link (`[Rng]`) to `https://rust-random.github.io/rand/rand_core/trait.RngCore.html`. This isn't quite right because it's an external link, but I definitely think this is the right approach. There are also some panics that look like I invalidated some assumptions in `resolve_str_path_error`:

```
(-bash@build-server) ~/.../src/test ▶️ RUST_BACKTRACE=1 rustdoc +stage1 rustdoc/issue-56701.rs
thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', src/librustc_resolve/build_reduced_graph.rs:111:17
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /home/joshua/.local/lib/cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /home/joshua/.local/lib/cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1076
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1537
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:218
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:486
  11: rust_begin_unwind
             at src/libstd/panicking.rs:388
  12: core::panicking::panic_fmt
             at src/libcore/panicking.rs:101
  13: core::panicking::panic
             at src/libcore/panicking.rs:56
  14: rustc_resolve::build_reduced_graph::<impl rustc_resolve::Resolver>::get_module
             at src/librustc_resolve/build_reduced_graph.rs:0
  15: rustc_resolve::Resolver::resolve_str_path_error
             at src/librustc_resolve/lib.rs:2925
  16: rustdoc::passes::collect_intra_doc_links::LinkCollector::resolve::{{closure}}
             at src/librustdoc/passes/collect_intra_doc_links.rs:147
  17: rustc_interface::passes::BoxedResolver::access::{{closure}}
             at /home/joshua/src/rust/src/librustc_data_structures/box_region.rs:113
  18: rustc_interface::passes::configure_and_expand::{{closure}}
             at src/librustc_interface/passes.rs:132
  19: alloc::boxed::<impl core::ops::generator::Generator<R> for core::pin::Pin<alloc::boxed::Box<G>>>::resume
             at /home/joshua/src/rust/src/liballoc/boxed.rs:1191
  20: rustc_data_structures::box_region::PinnedGenerator<I,A,R>::access
             at /home/joshua/src/rust/src/librustc_data_structures/box_region.rs:55
  21: rustc_interface::passes::BoxedResolver::access
             at /home/joshua/src/rust/src/librustc_data_structures/box_region.rs:119
  22: rustdoc::core::DocContext::enter_resolver
             at src/librustdoc/core.rs:76
  23: rustdoc::passes::collect_intra_doc_links::LinkCollector::resolve
             at src/librustdoc/passes/collect_intra_doc_links.rs:146
  24: <rustdoc::passes::collect_intra_doc_links::LinkCollector as rustdoc::fold::DocFolder>::fold_item
             at src/librustdoc/passes/collect_intra_doc_links.rs:627
```
@Manishearth is this because the module hasn't been serialized from the other crate into the .rlib? Or because of some other issue with `get_module` I don't know about?

**Comment by Manishearth** [comment]
@jyn514 That's fixing a different bug.

The bug we want to try and fix is when you have a reexport from a different crate and you do not put your own documenation on it, so the docs from the crate get used, and they get resolved relative to the local module instead of their source module.

I don't know why that panic is occurring, you'd need to know what item it is trying to resolve.

**Comment by jyn514** [comment]
@Manishearth I gave a bad example, but it fixes the other issue too. The original docs for `Rng` has a link to `RngCore` which now resolves after this PR.

**Comment by Manishearth** [comment]
@jyn514 this is still not the actual issue at hand. it's part of it, but the actual issue involves multiple crates, and the documentation being resolved needs to come from the _dependency_ crate. This is not about documentation on reexports itself.

Here are some test cases that should work:

```rust
//crate a

pub struct Bar;

/// Link to [Bar]
pub struct Foo;

// crate b
pub use a::Foo;
```

and the more complicated case of

```rust
// crate a

pub mod bar {
   pub struct Bar;
}

pub mod foo {
  use crate::bar;
  /// link to [bar::Bar]
  pub struct Foo;
}

// crate b

pub use a::foo::Foo;
```

Part of the issue is indeed fixing things so that reexports within a crate resolve correctly, but that's not all.
